File: app/src/main/java/de/jerleo/model/Meter.java
package de.jerleo.model;

import android.util.JsonReader;
import android.util.JsonWriter;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;

import de.jerleo.android.DateHelper;
import de.jerleo.database.Constants;
import de.jerleo.database.Database;


public class Meter {

    private final List<Reading> readings = new ArrayList<>();
    private final List<Tariff> tariffs = new ArrayList<>();

    private Home home;

    private long id;
    private String name;
    private String number;
    private Unit unit;

    private Meter prior;
    private long priorId;
    private String priorNumber;

    private HashMap<String, Integer> monthlyConsumption;

    public static List<Meter> importDatabase() {

        return Database.getMeterList();
    }

    public static List<Meter> importFrom(JsonReader reader) throws IOException {

        List<Meter> meters = new ArrayList<>();

        reader.beginArray();
        while (reader.hasNext())
            meters.add(Meter.importSingleFrom(reader));
        reader.endArray();

        return meters;
    }

    private static Meter importSingleFrom(JsonReader reader) throws IOException {

        Meter meter = new Meter();

        reader.beginObject();
        while (reader.hasNext()) {

            String nextName = reader.nextName();

            switch (nextName) {
                case Constants.NAME:
                    meter.setName(reader.nextString());
                    break;
                case Constants.NUMBER:
                    meter.setNumber(reader.nextString());
                    break;
                case Constants.UNIT:
                    meter.setUnit(Unit.valueOf(reader.nextString()));
                    break;
                case Constants.PRIOR:
                    meter.setPrior(reader.nextString());
                    break;
                case Constants.READINGS:
                    meter.setReadings(Reading.importFrom(reader));
                    break;
                case Constants.TARIFFS:
                    meter.setTariffs(Tariff.importFrom(reader));
                    break;
                default:
                    reader.skipValue();
                    break;
            }
        }
        reader.endObject();

        meter.update();
        return meter;
    }

    public void add(Reading reading) {

        reading.setMeter(this);
        reading.setPriceFrom(tariffs);
        reading.setPrior(getLastReading());
        reading.setId(Database.insert(reading));
        readings.add(0, reading);

        addCalculatedReadings();
        updateConsumption();
    }

    public void add(Tariff tariff) {

        tariff.setMeter(this);
        tariffs.add(tariff);
    }

    public void delete() {

        for (final Tariff tariff : tariffs)
            tariff.delete();
        tariffs.clear();

        for (final Reading reading : readings)
            reading.delete();
        readings.clear();

        Database.delete(this);
    }

    public void delete(Tariff tariff) {

        Database.delete(tariff);
        tariffs.remove(tariff);
        updateTariffValidity();
        updateConsumption();
    }

    public void deleteLastReading() {

        if (readings.size() == 0)
            return;

        final Reading reading = readings.get(0);
        final Reading prior = reading.getPrior();

        Database.delete(reading);
        readings.remove(0);

        if (prior != null && prior.isCalculated())
            readings.remove(prior);

    }

    public void export(JsonWriter writer) throws IOException {

        writer.beginObject();

        writer.name(Constants.NAME).value(name);
        writer.name(Constants.NUMBER).value(number);
        writer.name(Constants.UNIT).value(unit.name());

        if (hasPrior())
            writer.name(Constants.PRIOR).value(getPrior().getNumber());

        if (readings.size() > 0) {
            writer.name(Constants.READINGS);
            writer.beginArray();
            for (final Reading reading : readings)
                if (!reading.isCalculated())
                    reading.exportTo(writer);
            writer.endArray();
        }

        if (tariffs.size() > 0) {
            writer.name(Constants.TARIFFS);
            writer.beginArray();
            for (final Tariff tariff : tariffs)
                tariff.exportTo(writer);
            writer.endArray();
        }

        writer.endObject();
    }

    public float getAverageFor(final String month) {

        int sum = 0;
        int count = 0;

        for (String monthKey : monthlyConsumption.keySet())
            if (monthKey.substring(monthKey.length() - 2).equals(month)) {
                sum += monthlyConsumption.get(monthKey);
                count++;
            }

        if (hasPrior()) {
            HashMap<String, Integer> priorConsumption = getPrior().getMonthlyConsumption();
            for (String monthKey : priorConsumption.keySet())
                if (monthKey.substring(monthKey.length() - 2).equals(month)) {
                    sum += priorConsumption.get(monthKey);
                    count++;
                }
        }
        return count > 0 ? (float) sum / count : 0;
    }

    public int getConsumptionFor(Calendar month) {

        String monthKey = DateHelper.getMonthKey(month);

        return monthlyConsumption.containsKey(monthKey) ?
                monthlyConsumption.get(monthKey) : 0;
    }

    public float getCosts(Bill bill) {

        float costs = 0;
        final Calendar start = bill.getBegin();
        final Calendar end = bill.getEnd();
        for (final Reading r : readings)
            if (r.isBetween(start, end))
                costs += r.getCosts();
        return costs;
    }

    public float getFees(Bill bill) {

        final Calendar thisMonth = DateHelper.getToday();
        thisMonth.set(Calendar.DAY_OF_MONTH, 1);
        final Calendar firstMonth = bill.getBegin();
        final Calendar lastMonth = bill.getEnd().after(thisMonth) ?
                thisMonth : bill.getEnd();

        float totalFees = 0;
        while (lastMonth.compareTo(firstMonth) >= 0) {
            totalFees += getFeeForMonth(lastMonth);
            lastMonth.add(Calendar.MONTH, -1);
        }
        return totalFees;
    }

    public long getId() {

        return id;
    }

    public Reading getLastReading() {

        if (readings.size() > 0)
            return readings.get(0);
        else
            return null;
    }

    public String getName() {

        return name;
    }

    public String getNumber() {

        return number;
    }

    public float getPayments(Bill bill) {

        final Calendar thisMonth = DateHelper.getToday();
        thisMonth.set(Calendar.DAY_OF_MONTH, 1);
        final Calendar firstMonth = bill.getBegin();
        final Calendar lastMonth = bill.getEnd().after(thisMonth) ?
                thisMonth : bill.getEnd();

        float totalPayments = 0;
        while (lastMonth.compareTo(firstMonth) >= 0) {
            totalPayments += getPaymentForMonth(lastMonth);
            lastMonth.add(Calendar.MONTH, -1);
        }

        return totalPayments;
    }

    public Meter getPrior() {

        if (prior == null)
            if (priorId > 0)
                prior = home.getMeterById(priorId);
            else
                prior = home.getMeterByNumber(priorNumber);

        return prior;
    }

    public List<Reading> getReadings() {

        return readings;
    }

    public Tariff getTariff(int position) {

        return tariffs.get(position);
    }

    public List<Tariff> getTariffs() {

        return tariffs;
    }

    public Unit getUnit() {

        return unit;
    }

    public boolean hasPrior() {

        return getPrior() != null;
    }

    public void persist() {

        if (id == 0)
            id = Database.insert(this);
        else
            Database.update(this);

        for (Reading reading : readings)
            reading.persist();

        for (Tariff tariff : tariffs)
            tariff.persist();
    }

    public void setHome(Home home) {

        this.home = home;
    }

    public void setId(long id) {

        this.id = id;
    }

    public void setName(String name) {

        this.name = name;
    }

    public void setNumber(String number) {

        this.number = number;
    }

    public void setPrior(String number) {

        this.prior = null;
        this.priorId = 0;
        this.priorNumber = number;
    }

    public void setPrior(long id) {

        this.prior = null;
        this.priorNumber = null;
        this.priorId = id;
    }

    public void setReadings(List<Reading> newReadings) {

        readings.clear();
        readings.addAll(newReadings);
        addCalculatedReadings();
    }

    public void setTariffs(List<Tariff> newTariffs) {

        tariffs.clear();
        tariffs.addAll(newTariffs);
        for (final Tariff tariff : tariffs)
            tariff.setMeter(this);
    }

    public void setUnit(Unit unit) {

        this.unit = unit;
    }

    @Override
    public String toString() {

        return number + ";" + name + ";" + unit.name();
    }

    public void update() {

        updateTariffValidity();
        updateConsumption();
    }

    private void addCalculatedReadings() {

        List<Reading> calcReadings = new ArrayList<>();

        // Sort ascending by date
        Collections.sort(readings);

        // Initialize month
        int month = -1;

        // First reading has no priorNumber reading
        Reading prior = null;

        for (final Reading reading : readings) {

            // Get current month
            int currentMonth = reading.getDate().get(Calendar.MONTH);

            // Check for change of month ignoring first month
            if (month > -1 && month != currentMonth) {

                // Calculate interpolation
                Reading calculated = getInterpolation(reading, prior);

                // Check for insertion
                if (!calculated.equals(prior))
                    calcReadings.add(calculated);

                // Set priorNumber reading
                prior = calculated;
            }

            // Remember new month
            month = currentMonth;

            // Set reading attributes
            reading.setMeter(this);
            reading.setPrior(prior);

            // Set priorNumber reading for next iteration
            prior = reading;
        }

        // Add all calculated readings and sort descending
        readings.addAll(calcReadings);
        Collections.sort(readings, Collections.reverseOrder());
    }

    private float getFeeForMonth(Calendar month) {

        for (final Tariff tariff : tariffs)
            if (tariff.hasFee() && tariff.isValidFeeFor(month))
                return tariff.getFee();
        return 0;
    }

    private Reading getInterpolation(Reading current, Reading prior) {

        // Get reading dates
        Calendar currentDate = current.getDate();
        Calendar priorDate = prior.getDate();

        // Get end of last month
        Calendar lastMonth = DateHelper.getLastMonth(currentDate);

        // Prior reading was at end of last month
        if (lastMonth.equals(priorDate))
            return prior;

        // Calculate total days between readings
        int daysTotal = DateHelper.getDaysBetween(priorDate, currentDate);

        // Calculate days between current reading and end of month
        int daysToLast = DateHelper.getDaysBetween(lastMonth, currentDate);

        // Calculate interpolated meter count
        float usageTotal = current.getCount() - prior.getCount();
        float usagePerDay = usageTotal / daysTotal;
        float newCounter = current.getCount() - usagePerDay * daysToLast;

        // Create new interpolated reading
        Reading reading = new Reading();
        reading.setMeter(this);
        reading.setDate(lastMonth);
        reading.setPrior(prior);
        reading.setCount((int) newCounter);
        reading.setCalculated(true);

        return reading;
    }

    private HashMap<String, Integer> getMonthlyConsumption() {

        return monthlyConsumption;
    }

    private float getPaymentForMonth(Calendar month) {

        for (final Tariff tariff : tariffs)
            if (tariff.hasPayment() && tariff.isValidPaymentFor(month))
                return tariff.getPayment();
        return 0;
    }

    private void updateConsumption() {

        monthlyConsumption = new HashMap<>();

        for (final Reading reading : readings) {
            reading.setPriceFrom(tariffs);
            String month = reading.getMonth();
            int consumption = reading.getConsumption();

            if (monthlyConsumption.containsKey(month))
                consumption += monthlyConsumption.get(month);
            monthlyConsumption.put(month, consumption);
        }
    }

    private void updateTariffValidity() {

        Calendar validToFee = DateHelper.getToday();
        Calendar validToPrice = DateHelper.getToday();
        Calendar validToPayment = DateHelper.getToday();

        Collections.sort(tariffs, Collections.reverseOrder());

        for (final Tariff tariff : tariffs) {

            if (tariff.hasFee()) {
                tariff.setValidToFee((Calendar) validToFee.clone());
                validToFee = tariff.getValidFrom();
                validToFee.add(Calendar.DAY_OF_MONTH, -1);
            }

            if (tariff.hasPrice()) {
                tariff.setValidToPrice((Calendar) validToPrice.clone());
                validToPrice = tariff.getValidFrom();
                validToPrice.add(Calendar.DAY_OF_MONTH, -1);
            }

            if (tariff.hasPayment()) {
                tariff.setValidToPayment((Calendar) validToPayment.clone());
                validToPayment = tariff.getValidFrom();
                validToPayment.add(Calendar.DAY_OF_MONTH, -1);
            }
        }
    }

    public enum Unit {
        kWh, m3 {
            @Override
            public String toString() {

                return "m³";
            }
        }
    }
}

File: app/src/main/kotlin/de/jerleo/model/Meter.kt
package de.jerleo.model

import android.util.JsonReader
import android.util.JsonWriter
import de.jerleo.android.DateHelper
import de.jerleo.database.Constants
import de.jerleo.database.Database
import de.jerleo.database.Database.Companion.delete
import de.jerleo.database.Database.Companion.insert
import de.jerleo.database.Database.Companion.update
import java.io.IOException
import java.time.LocalDate

class Meter : Comparable<Meter> {

    lateinit var home: Home

    var name: String = ""
    var number: String = ""
    var unit: Unit = Unit.KWH

    var id: Long = 0L
    var priorId: Long = 0
    var priorNumber: String? = null
    var prior: Meter? = null
        get() {
            if (field == null)
                field = if (priorId > 0) home.meter(priorId) else
                    priorNumber?.let { home.meter(it) }
            return field
        }
        set(value) {
            if (field == value) return
            priorId = value?.id ?: 0
            priorNumber = value?.number
            field = value
        }

    var readings: MutableList<Reading> = ArrayList()
        set(value) {
            field.clear()
            field.addAll(value)
            updateReadings()
        }

    var tariffs: MutableList<Tariff> = ArrayList()
        set(value) {
            field.clear()
            field.addAll(value)
            updateTariffs()
        }

    override fun compareTo(other: Meter): Int = number.compareTo(other.number)
    override fun toString(): String = number + ";" + name + ";" + unit.name

    fun tariff(pos: Int): Tariff = tariffs[pos]
    fun tariff(date: LocalDate) = tariffs.firstOrNull { it.isValidPriceFor(date) }

    fun save(reading: Reading): Reading {
        val that = this
        readings.add(0,
            reading.apply {
                meter = that
                tariff = tariff(reading.date)
                prior = lastReading()
            })
        reading.persist()
        update()
        return reading
    }

    fun save(tariff: Tariff): Tariff {
        if (!tariffs.contains(tariff)) tariffs.add(tariff)
        tariff.persist()
        updateTariffs()
        update()
        return tariff
    }

    fun delete() {
        tariffs.forEach { it.delete() }
        tariffs.clear()
        readings.forEach { it.delete() }
        readings.clear()
        delete(this)
    }

    fun delete(tariff: Tariff) {
        Database.delete(tariff)
        tariffs.remove(tariff)
        updateTariffs()
    }

    fun deleteLastReading() {
        if (readings.isEmpty()) return
        val reading = readings.first()
        val prior = reading.prior
        delete(reading)
        readings.removeAt(0)
        if (prior != null && prior.isCalculated) readings.remove(prior)
    }

    @Throws(IOException::class)
    fun export(writer: JsonWriter) {
        writer.beginObject()
        writer.name(Constants.NAME).value(name)
        writer.name(Constants.NUMBER).value(number)
        writer.name(Constants.UNIT).value(unit.toString())
        prior?.let { writer.name(Constants.PRIOR).value(prior?.number) }
        if (tariffs.isNotEmpty()) {
            writer.name(Constants.TARIFFS)
            writer.beginArray()
            tariffs.forEach { it.exportTo(writer) }
            writer.endArray()
        }
        if (readings.isNotEmpty()) {
            writer.name(Constants.READINGS)
            writer.beginArray()
            readings.filter { !it.isCalculated }.forEach { it.exportTo(writer) }
            writer.endArray()
        }
        writer.endObject()
    }

    fun averageFor(month: Int): Float {
        var readings = readings.filter { it.date.monthValue == month }
        var months = readings.map { it.date.withDayOfMonth(1) }.distinct().count()
        var usage = readings.sumOf { it.usage() }.toFloat()
        prior?.let { meter ->
            readings = meter.readings.filter { it.date.monthValue == month }
            months += readings.map { it.date.withDayOfMonth(1) }.distinct().count()
            usage += readings.sumOf { it.usage() }.toFloat()
        }
        return if (months > 0) usage / months else 0f
    }

    fun usage(month: LocalDate) =
        readings.filter {
            it.date.year == month.year && it.date.monthValue == month.monthValue
        }.sumOf { it.usage() }

    fun payments(month: LocalDate): Double =
        tariffs.firstOrNull { it.isValidPaymentFor(month) }?.payment ?: 0.0

    fun costs(from: LocalDate, to: LocalDate): Double =
        readings.filter { it.date in from..to }.sumOf { it.costs() }

    fun fees(month: LocalDate): Double =
        tariffs.firstOrNull { it.isValidFeeFor(month) }?.fee ?: 0.0

    fun firstReading(): Reading? = readings.lastOrNull()
    fun lastReading(): Reading? = readings.firstOrNull()

    fun persist() {
        if (id == 0L) id = insert(this) else update(this)
        readings.forEach { it.persist() }
        tariffs.forEach { it.persist() }
    }

    fun update() {
        addCalculatedReadings()
        updateReadings()
    }

    private fun addCalculatedReadings() {
        val calculated: MutableList<Reading> = ArrayList()
        readings.filter { it.monthEndMissing() }.forEach { calculated.add(interpolated(it)) }
        readings.addAll(calculated)
    }

    private fun interpolated(reading: Reading): Reading {

        // Get reading dates
        val thisDate = reading.date
        val lastDate = reading.prior!!.date
        val endMonth = reading.date.withDayOfMonth(1).minusDays(1)

        // Calculate total days between readings
        val daysTotal = DateHelper.daysBetween(lastDate, thisDate)

        // Calculate days between current reading and end of month
        val daysToLast = DateHelper.daysBetween(endMonth, thisDate)

        // Calculate interpolated meter count
        val usageTotal = (reading.count - reading.prior!!.count).toFloat()
        val usagePerDay = usageTotal / daysTotal
        val newCounter = reading.count - usagePerDay * daysToLast

        // Get current meter
        val that = this

        // Return new interpolated reading
        return Reading().apply {
            meter = that
            date = endMonth
            count = newCounter.toInt()
            isCalculated = true
            tariff = tariff(date)
            prior = reading.prior
            reading.prior = this
        }
    }

    private fun updateTariffs() {
        var validToFee = DateHelper.today
        var validToPrice = DateHelper.today
        var validToPayment = DateHelper.today

        var lastFee = 0.0
        var lastPrice = 0.0
        var lastPayment = 0.0

        tariffs.sorted().forEach {
            if (it.hasFee()) lastFee = it.fee else it.lastFee = lastFee
            if (it.hasPrice()) lastPrice = it.price else it.lastPrice = lastPrice
            if (it.hasPayment()) lastPayment = it.payment else it.lastPayment = lastPayment
        }
        tariffs.sortDescending()
        tariffs.forEach {
            if (it.hasFee()) {
                it.dateFee = validToFee
                validToFee = it.dateFrom.minusDays(1)
            }
            if (it.hasPrice()) {
                it.datePrice = validToPrice
                validToPrice = it.dateFrom.minusDays(1)
            }
            if (it.hasPayment()) {
                it.datePayment = validToPayment
                validToPayment = it.dateFrom.minusDays(1)
            }
            it.meter = this
        }
    }

    private fun updateReadings() {
        var prior: Reading? = null
        readings.sort()
        readings.forEach {
            it.meter = this
            it.tariff = tariff(it.date)
            it.prior = prior
            prior = it
        }
        readings.sortDescending()
    }

    enum class Unit {
        KWH {
            override fun toString(): String = "kWh"
        },
        M3 {
            override fun toString(): String = "m³"
        }
    }

    companion object {

        val import by lazy { Database.meters }

        @Throws(IOException::class)
        fun import(reader: JsonReader): MutableList<Meter> {
            val meters: MutableList<Meter> = ArrayList()
            reader.beginArray()
            while (reader.hasNext()) meters.add(importSingle(reader))
            reader.endArray()
            return meters
        }

        @Throws(IOException::class)
        private fun importSingle(reader: JsonReader): Meter {
            val meter = Meter()
            reader.beginObject()
            while (reader.hasNext()) {
                when (reader.nextName()) {
                    Constants.NAME -> meter.name = reader.nextString()
                    Constants.NUMBER -> meter.number = reader.nextString()
                    Constants.UNIT -> {
                        val unitString = reader.nextString()
                        meter.unit = Unit.values().find { unitString == it.toString() } ?: Unit.M3
                    }
                    Constants.PRIOR -> meter.priorNumber = reader.nextString()
                    Constants.TARIFFS -> meter.tariffs = Tariff.import(reader)
                    Constants.READINGS -> meter.readings = Reading.import(reader)
                    else -> reader.skipValue()
                }
            }
            reader.endObject()
            meter.update()
            return meter
        }
    }
}